theseus/optimizer/nonlinear/levenberg_marquardt.py

```diff
@@ -54,31 +73,87 @@ def __init__(
+    ) -> None:
         if ellipsoidal_damping and not self._allows_ellipsoidal:
             raise NotImplementedError(
                 f"Ellipsoidal damping is only supported by solvers with type "
-                f"[{solvers_str}]."
+                f"[{_EDAMP_SOLVERS_STR}]."
+            )
+        self._ellipsoidal_damping = ellipsoidal_damping
+        if adaptive_damping and not self._allows_adaptive:
+            raise NotImplementedError(
+                f"Adaptive damping is only supported by solvers with type "
+                f"[{_ADAPT_DAMP_SOLVERS_STR}]."
+            )
+        if adaptive_damping:
+            self._damping = damping * torch.ones(
+                self.objective.batch_size,
+                device=self.objective.device,
+                dtype=self.objective.dtype,
             )
-        if damping_eps and not self._allows_ellipsoidal:
+        else:
+            self._damping = damping
+
+    def compute_delta(
+        self,
+        damping_eps: Optional[float] = None,
+        **kwargs,
+    ) -> torch.Tensor:
+        if damping_eps is not None and not self._allows_ellipsoidal:
             raise NotImplementedError(
                 f"damping eps is only supported by solvers with type "
-                f"[{solvers_str}]."
+                f"[{_EDAMP_SOLVERS_STR}]."
             )
-        damping_eps = damping_eps or 1e-8
+        damping_eps = damping_eps if damping_eps is not None else 1e-8
 
         return self.linear_solver.solve(
-            damping=damping,
-            ellipsoidal_damping=ellipsoidal_damping,
+            damping=self._damping,
+            ellipsoidal_damping=self._ellipsoidal_damping,
             damping_eps=damping_eps,
         )
+
+    # Updates damping per batch element depending on whether the last step
+    # was successful in decreasing error or not.
+    # Based on https://people.duke.edu/~hpgavin/ce281/lm.pdf, Section 4.1
+    # We currently use method (1) from 4.1.1
+    def _update_state_impl(
+        self, last_err: torch.Tensor, new_err: torch.Tensor, delta: torch.Tensor
+    ) -> None:
+        if not self._adaptive_damping:
+            return
+        linearization = self.linear_solver.linearization
+        if not isinstance(linearization, DenseLinearization):
+            warnings.warn(
+                "Adaptive damping is currently only supported with "
+                "DenseLinearization. Damping will not update.",
+                RuntimeWarning,
+            )
+            return
+
+        damping = (
+            self._damping.view(-1, 1)
+            if isinstance(self._damping, torch.Tensor)
+            else self._damping
+        )
+        den = (delta * (damping * delta + linearization.Atb.squeeze(2))).sum(dim=1)
+        rho = (last_err - new_err) / den
+        good_idx = rho > self._ACCEPT_EPS
+        self._damping = torch.where(good_idx, self._damping / 9, self._damping * 11)
+        self._damping = self._damping.clamp(1.0e-7, 1.0e7)
```

Comment:
Are 9, 11, 1e-7, 1e7 fixed numbers of the algorithm or hyperparameters? If latter then maybe we shouldn't hardcode them.

Reply:
Ah looking from the doc they are fixed numbers in the algorithm for 'good' convergence.

Reply:
Yeah, but also not a bad idea to make them named constants instead of magic numbers.

Reply:
Agreed. Then we can easily edit them in the future if needed. Potentially add them to constants.py.

Reply:
I think we should make these configurable parameters as some applications uses different values. 
For example GTSAM and ceres uses different parameters (as ceres is tuned more for SfM)
https://github.com/borglab/gtsam/blob/develop/gtsam/nonlinear/LevenbergMarquardtParams.h#L85

Reply:
Good point @rmurai0610, then we need these to be kwargs rather than in constants.py.

Reply:
Good idea. Added!